proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,186 @@
+- primary_retriever: First retriever.
+- secondary_retriever (optional): Second retriever (if applicable).
+- keep_original_score (optional): Set this to store the original relevance score of the returned document in the document's meta field (the documents' scores get replaced in the FileSimilarityRetriever output with the reciprocal rank fusion score).
+- top_k: How many documents to return.
+- max_query_len: How many characters can be in a query document. The documents exceeding this limit will be cut off. This was added because in some cases BM25Retriever threw an error if the query exceeded the `max_clause_count` search setting (https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-settings.html)
+- max_num_queries (optional): The maximum number of queries that should be run for a single file. If the number of query documents exceeds this limit, the query documents will be split into n parts so that n < max_num_queries and every nth document will be kept.
+- use_existing_embeddings: Whether to re-use the existing embeddings from the index. To optimize speed for the file similarity retrieval you should set this parameter to `True`. This way the FileSimilarityRetriever can run on the CPU.
+
+# Detailed design - updated
+
+During the proposal discussion, it was decided to break up the initial FileSimilarityRetriever into two sub-components in order to simplify the design and testing and also to make it possible to use only one of the sub-components, if needed, thus covering more use cases.
+
+The new design would go as follows:
+- DocumentFetcher that gets all docs corresponding to a file based on a provided aggregation meta key and returns a list of docs to be used by one or two retrievers as input for batch retrieval.
+- DocumentAggregator placed after the retriever(s) that aggregates the document list(s) using RRF into a single ranked list of files, which are put together based on file_ids of retrieved docs.
+
+These sub-components will have the following params (for full explanation see Detailed design section above):
+- DocumentFetcher
+  - document_store
+  - file_aggregation_key
+  - max_num_docs: optional, in the original version it was called max_num_queries
+  Note: When researching how to change the params from the original implementation I found out there is no longer need for max_query_len that was required in the original FileSimilarityRetriever. It was implemented to not overshoot the max_clause_count limit in ElasticSearch settings but this got deprecated as of vs 8.0.0 according to their documentation - https://www.elastic.co/guide/en/elasticsearch/reference/current/search-settings.html
+
+- DocumentAggregator
+  - documents: can be a list or a list of list of docs
+  - file_aggregation_key: default would be "file_id", but can be changed to e.g. "name"
+  - output: default "top_document", but can also be "file_aggregation_key"
```

Comment:
Hmm, given the types of output that we plan on supporting I think we should find a different name than DocumentAggregator. It feels like a too general of a name that a user in isolation might want to use for other purposes. And given it's name I would expect the output to be a List of List of Documents grouped together based on the `file_aggregation_key`. 

Actually maybe we can keep the name, but I think we should also support the List of List of Documents output as well.

Reply:
Ah, here "output" refers to the parameter named "output" which we've [discussed earlier](https://github.com/deepset-ai/haystack/pull/5629/files#diff-5c6cac19742c9d58b46f9d0debaf53f941d926db1ef5e5cea0e638222619f4efR125) (let me know if it's confusing, can rename if needed) rather than the format of the output itself.